### ISSUE TITLE
Underlying webdriver instance does not support javascript

### ISSUE THREAD

**Issue by DavidSelby** [appium/appium#5623]
1.5-beta (can't add labels)

When calling `@driver.execute_script("window.localStorage.clear();")` the following error is returned: "underlying webdriver instance does not support javascript".

Driver setup:

`caps = {`
        `'platformName' => 'iOS',`
        `'platformVersion' => '8.4',`
        `'deviceName' => 'iPhone Simulator',`
        `'browserName' => 'Safari'`
   `}`
   `$appium = Appium::Driver.new(:caps => caps, :appium_lib => {'port' => '4723')`
    `$driver = $appium.start_driver`

**Comment by JaniJegoroff** [comment]
I see the same issue with `v1.5.0-beta11 (REV 300441be033c7e47ae26de29137aa985ad7169d4)`. Appium server `1.4.16` is working fine.

```
Janis-MacBook-Air:JLPermissions janijegoroff$ ruby test.rb
/Users/janijegoroff/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/appium_lib-8.0.0/lib/appium_lib/common/wait.rb:50:in `_generic_wait': timed out after 30 seconds (underlying webdriver instance does not support javascript) (Selenium::WebDriver::Error::TimeOutError)
	from /Users/janijegoroff/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/appium_lib-8.0.0/lib/appium_lib/common/wait.rb:95:in `wait'
	from /Users/janijegoroff/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/appium_lib-8.0.0/lib/appium_lib/driver.rb:229:in `rescue in block (4 levels) in promote_appium_methods'
	from /Users/janijegoroff/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/appium_lib-8.0.0/lib/appium_lib/driver.rb:220:in `block (4 levels) in promote_appium_methods'
	from test.rb:18:in `<main>'
Janis-MacBook-Air:JLPermissions janijegoroff$
```

Here is the simple example project to reproduce: https://github.com/JaniJegoroff/JLPermissions

**Comment by scottdixon** [comment]
@imurchie has made some fixes around execute_script, these should be published in the next release in the coming days.

**Comment by JaniJegoroff** [comment]
@scottdixon @imurchie Any updates on this? I'm currently blocked with iOS 10 beta release due to this issue.

**Comment by imurchie** [comment]
@JaniJegoroff Can you provide the server logs when this fails?

**Comment by JaniJegoroff** [comment]
@imurchie Here is the server log: https://gist.github.com/JaniJegoroff/07928874c5477d5f5548cce00ac1a439

```
Scenario: show completed swipe offer details                   # features/swipe_offers.feature:7
    Given server returns completed swipe offer details           # features/step_definitions/swipe_offers_steps.rb:5
    And I am on wallet offers screen                             # features/step_definitions/common_steps.rb:1
    When I select wallet offer                                   # features/step_definitions/wallet_steps.rb:31
    Then I should see congratulations for completing swipe offer # features/step_definitions/swipe_offers_steps.rb:22
      timed out after 30 seconds (underlying webdriver instance does not support javascript) (Selenium::WebDriver::Error::TimeOutError)
      ./features/step_definitions/swipe_offers_steps.rb:26:in `/^I should see congratulations for completing swipe offer$/'
      features/swipe_offers.feature:11:in `Then I should see congratulations for completing swipe offer'
```

For some reason I got a feeling I don't have the latest Appium beta release installed even the log says `[Appium] Welcome to Appium v1.6.0-beta1 (REV b77a695b7a0c9094d49993d26921a47da919364d)`.

**Comment by JaniJegoroff** [comment]
@imurchie I rerun `npm install -g appium@beta` and noticed quite a many npm package updates. However Appium server REV code remains the same.

Now the error is different:
https://gist.github.com/JaniJegoroff/f660d67be26a170ac0ae29b7230ce7e2

```
Scenario: show completed swipe offer details                   # features/swipe_offers.feature:7
    Given server returns completed swipe offer details           # features/step_definitions/swipe_offers_steps.rb:5
    And I am on wallet offers screen                             # features/step_definitions/common_steps.rb:1
    When I select wallet offer                                   # features/step_definitions/wallet_steps.rb:31
    Then I should see congratulations for completing swipe offer # features/step_definitions/swipe_offers_steps.rb:22
      timed out after 30 seconds (Method is not implemented) (Selenium::WebDriver::Error::TimeOutError)
      ./features/step_definitions/swipe_offers_steps.rb:26:in `/^I should see congratulations for completing swipe offer$/'
      features/swipe_offers.feature:11:in `Then I should see congratulations for completing swipe offer'
```

**Comment by imurchie** [comment]
Yes, the main Appium package hasn't changed but we are working hard on the XCUITest package, so it gets updated a lot.

The script you are trying to execute is UI Automation code? In which case it will _not_ work, no matter what, since UI Automation does not exist anymore.

**Comment by JaniJegoroff** [comment]
I'm using Appium Ruby binding methods `text` and `button` (not a custom UIA script):
https://github.com/appium/ruby_lib/blob/master/lib/appium_lib/ios/element/text.rb#L10
https://github.com/appium/ruby_lib/blob/master/lib/appium_lib/ios/element/button.rb#L10

Eventually execution comes to the iOS Helper module:
https://github.com/appium/ruby_lib/blob/master/lib/appium_lib/ios/helper.rb#L520

I see, helper file contains some raw UIA code. But not sure what happens under the hood in this case, especially when execution comes to https://github.com/appium/ruby_lib/blob/master/lib/appium_lib/ios/helper.rb#L546

**Comment by jlipps** [comment]
Sounds like the Ruby client needs to be updated to throw an error if someone tries to use any of these UIAutomation-based helper methods on Xcode 8+.

**Comment by JaniJegoroff** [comment]
Thanks guys, I'm going to refactor my Ruby tests using alternative methods.

**Comment by danqa** [comment]
Hey guys.

I am getting this error as well:

```
Selenium::WebDriver::Error::UnsupportedOperationError:
       underlying webdriver instance does not support javascript
```

when I try to do 
$selenium_driver.execute_script('return document.querySelectorAll(".readonly").length')

not using a UIA call. Any idea what might cause that? Is this a ruby client issue?

Thanks in advance.

**Comment by jlipps** [comment]
The ruby client might be checking the create session response to see if it includes the javascript support key?

**Comment by danqa** [comment]
I added :javascript_enabled => true but still getting that error.

**Comment by jlipps** [comment]
@danqa can you please provide a gist of your appium logs?

**Comment by danqa** [comment]
@jlipps  https://gist.github.com/danqa/52797e17e6c1a8f784ad05af6ec20439

Here is the failure:

```
 Sessions Quick Taps verifies quickly tapping the footer buttons does not lock up the app
     Failure/Error:
       $appium_driver.wait { 
         $selenium_driver.execute_script('return document.querySelector(".info__nav__desc")') != [] 
       }

     Selenium::WebDriver::Error::TimeOutError:
       timed out after 30 seconds (underlying webdriver instance does not support javascript)
```

**Comment by jlipps** [comment]
@danqa these are the capabilities your script is sending in:

```
312016-11-09 23:03:07:899 - [Appium] Capabilities:
322016-11-09 23:03:07:901 - [Appium] deviceName: 'iPhone Simulator'
332016-11-09 23:03:07:901 - [Appium] automationName: 'XCUITest'
342016-11-09 23:03:07:902 - [Appium] webdriver.remote.quietExceptions: true
352016-11-09 23:03:07:902 - [Appium] autoWebview: 'true'
362016-11-09 23:03:07:902 - [Appium] sendKeyStrategy: 'setValue'
372016-11-09 23:03:07:902 - [Appium] javascript_enabled: true
382016-11-09 23:03:07:903 - [Appium] platformVersion: '10.0'
392016-11-09 23:03:07:903 - [Appium] fullReset: 'true'
402016-11-09 23:03:07:904 - [Appium] platformName: 'iOS'
412016-11-09 23:03:07:904 - [Appium] orientation: 'PORTRAIT'
422016-11-09 23:03:07:905 - [Appium] launchTimeout: 180000
432016-11-09 23:03:07:905 - [Appium] showIOSLog: true
442016-11-09 23:03:07:906 - [Appium] noReset: true
452016-11-09 23:03:07:906 - [Appium] backendRetries: 4
462016-11-09 23:03:07:906 - [Appium] newCommandTimeout: 0
472016-11-09 23:03:07:906 - [Appium] app: '/tmp/tmpyL9Fa_/X_Operator.app'
482016-11-09 23:03:07:907 - [Appium] keepKeyChains: true
```

try using `javascriptEnabled` instead of `javascript_enabled`

and once again @imurchie it would be great to have less truncation of basedriver response here :-(

**Comment by danqa** [comment]
Hi @jlipps 

I tried that and it didn't work. I spoke to the folks at sauce as its happing in their cloud but not for me locally (which is very odd). They could not reproduce using Java but when they tried with appium_lib ruby bindings they can reproduce. So it seems to be an issue with appium_lib. I opened a ticket with them: https://github.com/appium/ruby_lib/issues/390 to hopefully get this addressed. 

Thank you for your help.

Sincerely
